### ISSUE TITLE
Shield immediately in coinbase

### ISSUE THREAD

**Issue by ebfull** [zcash/zcash#2488]
If we follow #2171 and the general approach of the Sapling design that Ian/Matt/I are working on, it will be more than feasible to "mint" (as in Zerocash) in the coinbase transaction, allowing us to enforce that miner rewards (and the founder's reward) is shielded immediately.

**Comment by nathan-at-least** [comment]
Consider how this relates to this other feature change proposal, #2170 *Disunify Mint and Pour*, and especially the use case I describe [in this comment](https://github.com/zcash/zcash/issues/2170#issuecomment-299057013).

I want to ensure we support the miner payout use case as well as possible (with maximal privacy which still supports the user behavior).

With this ticket implemented, could a coinbase do a payout to hundreds of pool participants? Do existing pool protocols (esp. p2pool) support this, or can they be modified to support this? (Paging @str4d for p2pool expertise.)

**Comment by ebfull** [comment]
Yeah, the actual act of creating the notes in the coinbases should be very cheap, so it should be able to handle hundreds of pool participants.

**Comment by str4d** [comment]
Existing pool protocols don't have the miners creating the coinbase transaction; the mining pool does. So they can pay to whoever they want.

P2Pool would be an interesting case, because miners create the coinbases themselves. In its transparent form, a miner creates a coinbase with its own address first, followed by the first addresses in the previous K blocks in the P2Pool chain (this is trivially equivalent to taking the list in the previous block, dropping the last address, and prepending their address - that just doesn't validate the block against the P2Pool chain).

If we shield immediately to coinbase, the target address would obviously be hidden. So a miner would not be able to construct a coinbase transaction that pays to previous addresses solely using the coinbase outputs of previous transactions. There are two directions this could go:

- Transparent coinbase outputs can be cut-and-pasted into new blocks. Is it possible for shielded coinbase outputs to similarly be moved around, or would that invalidate the proofs?
  - If this was possible, the P2Pool chain consensus rule would be that a new chain block must contain the first shielded coinbase proof from each of the last K blocks.
- Store the full list of addresses somewhere else (note: not set, because addresses appearing more than once correspond to a miner having more "shares"). This would likely require modifying the P2Pool protocol to send the list of addresses alongside the blocks, and committing to them somehow (e.g. hash them together, and then store the result alongside the existing P2Pool block chaining hash).
  - There would need to be some way to prove that the given shielded coinbase outputs corresponded to the list of transparent addresses.
  - The transparent addresses would not be stored in the Zcash blockchain, but the attestation would remain as for regular P2Pool blocks.
  - I don't remember right now where P2Pool blocks store their chaining information, but it is likely inside the coinbase outputs. We'd need to ensure that this was still possible somehow (as this data *does* need to end up on the Zcash blockchain).

**Comment by ebfull** [comment]
> If we shield immediately to coinbase, the target address would obviously be hidden.

This isn't actually true unless we use a zk-SNARK, though it may be extremely inexpensive to do so with the new commitment format...

**Comment by str4d** [comment]
@ebfull okay, then I need to review the proposed disunification. If having shielded addresses as coinbase outputs does *not* hide the target address, then P2Pool should operate fine (as long as it remains trivial to validate that the outputs do in fact correspond to those addresses, ie. the addresses are listed). But OTOH, that's not great for privacy, as it would enable those notes to be trivially eliminated from the global anonymity set?

**Comment by ebfull** [comment]
> as it would enable those notes to be trivially eliminated from the global anonymity set?

No, they would still be hidden. Nobody knows when they're spent. They can just see the destination address, which is not ideal since the user may be reusing it elsewhere.

**Comment by daira** [comment]
A Mint in Zerocash does hide the destination address; it only reveals the value and the commitment. A Mint/Shield in Sapling would be the same, except that it also needs to encrypt the coin to its creator's address in order to support recovering coins given only a spending key (#32).

**Comment by ebfull** [comment]
Yeah, I was wrong. I was picturing our use of pedersen commitments not having that kind of layering and instinctively thought it wouldn't be secure to support cheap minting.

**Comment by daira** [comment]
The first comment suggests also shielding Founders' Reward immediately, which is a separable decision. I don't think there is a good reason to do that; it would require changing all the future Founders' Reward addresses, and the code enforcing that the Founders' Reward is paid. That seems like make-work, and it might prompt FUD about us having something to hide (which would be without basis since for operational security reasons the Founders' Reward is already moved from the original addresses relatively quickly, but why invite more FUD?)

**Comment by str4d** [comment]
@jackgavigan raised a good point to me: if we shield coinbase immediately, and the addresses *are* hidden, then we lose the ability to monitor [miner distribution and mining pool sizes](https://explorer.zcha.in/statistics/miners), which seems like it would be significantly detrimental. It seems to me like it would make sense to develop a Mint where both the value *and* address are kept public for accountability, but where the generated note can be spent in a shielded transaction for unlinkability.

**Comment by ebfull** [comment]
In the coinbase, we can do the minting in a way that publishes the outgoing address.

**Comment by daira** [comment]
@str4d wrote:
>If we shield immediately to coinbase, the target address would obviously be hidden. So a miner would not be able to construct a coinbase transaction that pays to previous addresses solely using the coinbase outputs of previous transactions. [...]
>* Transparent coinbase outputs can be cut-and-pasted into new blocks. Is it possible for shielded coinbase outputs to similarly be moved around, or would that invalidate the proofs?

Yes, the inner note commitment from the coinbase output acts as a capability to pay a given address, without revealing the address. So it can be moved around in the way needed.

Remember that pools only reuse transparent addresses by choice. They could just as easily use a unique address for each block (and I'm mildly surprised that they don't).

**Comment by daira** [comment]
From https://github.com/zcash/zcash/issues/2170#issuecomment-299057013 :
> Currently one or more mining pool operators have reported that they cannot do shielded payouts, even when they or their miners would prefer to.
>
> We should verify with pool operator use case that this would allow payouts to arrive at z-addresses. IIUC, this would look like a single transaction with an aggregation of coinbase mining rewards with a very large Shield/Mint vector, with individual public values as payouts. This means the distribution of values would be non-private. The benefit however is that recipients would receive payouts to a z-address, and if any of them postpone the time until they deshield or they send those through subsequent shielded transfers, they would be improving system-wide privacy.

**Comment by daira** [comment]
Huh, I don't remember this having been rejected. Keeping the current "coinbase transactions can't have shielded outputs but their outputs are special and must then be shielded" behaviour for Sapling would be really ugly.

I think that with the performance improvement for shielded outputs especially (which only require a proof for a circuit with less than 2^14 constraints in Sapling), there's no longer any justification for this complication, which leads to much confusion and gnashing of teeth in practice.

**Comment by daira** [comment]
I've been persuaded that we should keep the current rules for Sapling, with Spend and Output descriptions as well as JoinSplit descriptions prohibited from coinbase transactions (see https://github.com/zcash/zcash/pull/567#commitcomment-15254012 for the original reasoning). We'll change it later.

**Comment by daira** [comment]
I think this is a strong contender for Blossom. If we change the FR addresses ([as already planned](https://forum.zcashcommunity.com/t/announcing-zcash-blossom-and-proposed-feature-goals/31891)), we can also make those shielded.

**Comment by daira** [comment]
I continue to believe this would be a substantial improvement, and I'm somewhat disappointed that it didn't get more support in the feature selection meeting. Can the reasons for deferring it be documented, please?

**Comment by daira** [comment]
This does not require a circuit change.

**Comment by ebfull** [comment]
We need to carefully understand the places in our codebase where coinbase transactions are treated differently from regular transactions -- in terms of how they can or do impact the UTXO set in various ways. Annoyingly, upstream bitcoin makes many assumptions and these originally led to how our mandatory-shielding policies were implemented. We may have made things worse by following their lead with more subtle invariants. :(

I think it would be worthwhile to explore this in-depth and then come up with a draft ZIP for making immediate shielding mandatory for mining rewards.

**Comment by nathan-at-least** [comment]
Note, regardless of whether or not #3677 is completed, let's specify this to only support Sapling outputs as a nudge away from Sprout addresses.

**Comment by str4d** [comment]
Note that the funding stream addresses defined in ZIP 207 (to be deployed in Blossom) assume transparent outputs in the coinbase, and changing those addresses is likely to require more overhead than we have time for in NU3. So we at least need to allow funding stream outputs to be transparent. That means we could implement this in two ways:

- Require funding stream outputs to be transparent, and all other outputs to be shielded, in NU3. Later, switch all funding streams to be shielded.
- Allow outputs to be either transparent or shielded in NU3, and then in a later NU require all outputs to be shielded.

@ebfull and I think that the second option is the better option implementation-wise, and it gives mining pools etc. time to migrate.

In order to still enforce that funding streams are paid to the correct addresses in the correct amounts, we would add a consensus rule that Sapling outputs in the coinbase transaction **MUST** use a fixed `ovk` - probably the all-zeroes `ovk`. Then the funding stream consensus rules would decrypt every Sapling output in the consensus rules in order to find the funding stream outputs.

We would additionally require that all Sapling outputs in the coinbase transaction **MUST** have valid commitments. This means we aren't coupling the use of AEAD for the ciphertexts (which was being considered for removal in #3226) to this consensus rule; we are just requiring that whatever the output of ciphertext decryption is, it correctly corresponds to the note that the proof was computed over. This rule implies that target addresses and values for _all_ Sapling outputs within the coinbase are revealed, which maintains the ability for network observers to track miners and mining pools. Meanwhile, the miners and mining pools could put useful / identifying text in the memo field, instead of stuffing it elsewhere in the coinbase transaction.

**Comment by str4d** [comment]
More thoughts from my pairing with @ebfull:

- The current 100-block coinbase maturity won't work as it currently does, because by design we can't tell when a Sapling output gets spent! We could resolve this in two ways:
  - Don't enforce coinbase maturity for Sapling coinbase outputs.
    - Is this a problem?
  - Add the commitments for Sapling coinbase outputs to the Sapling commitment tree 100 blocks after the block is mined.
    - This could get tricky implementation-wise.
  - Add this back in using a limited form of programmability.
    - This would require a circuit change.
- `CMerkleTx::GetBlocksToMaturity()` is already checked in the wallet when selecting shielded notes. If we don't enforce shielded coinbase maturity in the consensus rules, and also don't change this function, we'd end up with a standard rule where `zcashd` doesn't spend shielded coinbase until the maturity passes, but where technically it would be possible to do so.
- Enforcing all Sapling outputs to be valid, and thus revealing miner addresses, enables future mining consensus rule changes that require e.g. being able to pay mining rewards to prior miners.
- We need to split out the Sapling output checks from `CCoinsView::HaveShieldedRequirements()`, and move the Sapling output checks outside the `!tx.IsCoinbase()` check in `ConnectBlock()`.
- We need to document that the default zcashd behaviour for selecting a shielded mining address will differ from transparent address selection. Transparent addresses are selected fresh for every block (at least after #3762 was merged, which fixed a long-standing bug here), while we would just use a single address for shielded mining (matching our existing messaging and UX around treating addresses as buckets of money, instead of the entire wallet).
- We will need to modify `IncrementExtraNonce()` to _not_ modify the coinbase, because that will break `bindingSig`.

**Comment by str4d** [comment]
Closed by #4256.
